Improves validation against JUnit schema

yamokosk opened the pull request:
The two functions that generated JUnit XML were missing some required
attributes as defined by the JUnit schema:

https://raw.githubusercontent.com/windyroad/JUnit-Schema/master/JUnit.xsd

These missing attributes were messing with downstream tools that expected
those attributes to be present.

Also values for other XML attributes were improved by using standard Python
modules (datetime and socket).

yamokosk committed:
Improves validation against JUnit schema

The two functions that generated JUnit XML were missing some required
attributes as defined by the JUnit schema:

https://raw.githubusercontent.com/windyroad/JUnit-Schema/master/JUnit.xsd

This missing attributes were messing with downstream tools that expected
those attributes to be present.

yamokosk committed:
Fixes import statement order (flake8)

yamokosk committed:
Fixes import order (again)

johnthagen commented:
@yamokosk Thanks for the PR.

This script was originally written for [Bamboo](https://www.atlassian.com/software/bamboo), I'm curious what you are testing against to have an idea how this will improve compatibility.

I feel like we should update the unit tests too, but `hostname()` and `timestamp` will be hard to validate statically. Any ideas?

yamokosk commented:
I'm certainly no expert with schemas, but as far as I can tell the attributes I added are required per the schema (e.g. ['errors'](https://github.com/windyroad/JUnit-Schema/blob/master/JUnit.xsd#L190)). But it is also driven by interoperability. In this case, with ROS's [catkin](https://github.com/ros/catkin), and specifically with this [module](https://github.com/ros/catkin/blob/kinetic-devel/python/catkin/test_results.py).

As far as testing? I think something quick could be a nosetest that uses the class to write a dummy JUnit file and then drops to console and simply runs something like,

```
xmllint --noout --schema https://raw.githubusercontent.com/windyroad/JUnit-Schema/master/JUnit.xsd ./test.xml
```

I think you were using a pretty standard Python XML module and most everything else was standard Python modules. That should be a sufficient test.

I could, of course, be all wrong.. like I said, I really don't know much when it comes to XSDs.

johnthagen commented:
Okay, yeah I was thinking about just updating the [test.py](https://github.com/johnthagen/cppcheck-junit/blob/master/test.py) in this repo to check for the existence of the new `dict` entries you added so that they are tested in the future. I think that should be all that's needed to merge this.

Thanks.

johnthagen reviewed (changes requested):
Need to update test.py to check that 'timestamp', and 'hostname' attributes are created.

yamokosk committed:
Brings generated JUnit XML inline with spec

Added code to test cases to check that all required attributes for
the testsuite, testcase, and error tags are present in the XML
generated by this module.

Also had to add some new attributes to make new unittests pass.

yamokosk committed:
Fixes line length

yamokosk commented:
Done.

johnthagen commented:
Thanks!

johnthagen commented:
@yamokosk  Please try the [1.4.0 release](https://pypi.python.org/pypi/cppcheck-junit/1.4.0).
